File: go/fib.go
package main

import (
  "os"
  "fmt"
  "strconv"
  "math"
  "math/big"
)

func fib(n int) int {
  if n < 2 {
    return n
  }
  return fib(n - 1) + fib(n - 2)
}

func fibLinear(n int) int {
  if n < 2 {
    return n
  }
  prevFib, fib := 0, 1
  for ;n > 1; n-- {
    fib, prevFib = fib + prevFib, fib
  }
  return fib
}

func fibFormula(n int) int {
  return int(math.Round(math.Pow((math.Pow(5, .5) + 1) / 2, float64(n)) / math.Pow(5, .5)))
}

func fibTailAuxiliary(n int, prevFib int, fib int) int {
  if n == 0 {
    return prevFib
  }
  return fibTailAuxiliary(n - 1, fib, prevFib + fib)
}
func fibTailRecursive(n int) int {
  return fibTailAuxiliary(n, 0, 1)
}

func fibFastAuxiliary(n int) (*big.Int, *big.Int) {
  if n == 0 {
    return big.NewInt(0), big.NewInt(1)
  }
  prevFib, fib := fibFastAuxiliary((n / 2) | 0)
  temp := big.NewInt(0).Mul(fib, big.NewInt(2))
  temp.Sub(temp, prevFib)
  temp.Mul(temp, prevFib)

  prevFib.Mul(prevFib, prevFib)
  fib.Mul(fib, fib)
  fib.Add(fib, prevFib)

  if n % 2 == 0 {
    return temp, fib
  }
  return fib, temp.Add(temp, fib)
}

func fibFastDoubling(n int) *big.Int {
  result, _ := fibFastAuxiliary(n)
  return result
}

func fibLinearBig(n int) *big.Int {
  prevFib, fib := big.NewInt(0), big.NewInt(1)
  for ;n > 1; n-- {
    fib, prevFib = big.NewInt(0).Add(fib, prevFib), fib
  }
  return fib
}

func main() {
  input := 29
  if len(os.Args) > 1 {
    conversion, err := strconv.Atoi(os.Args[1])
    if err == nil {
      input = conversion
    }
  }
  fibonacci := fibFastDoubling(input)
  fmt.Printf("Fibonacci %d: %d\n", input, fibonacci)
}

File: cplusplus/fib.cpp
#include <iostream>
#include <math.h>

int fib(int n)
{
  return n < 2 ? n : fib(n - 1) + fib(n - 2);
}

int fibLinear(int n, int prevFib = 0, int fib = 1)
{
  if (n < 2) return n;
  while (--n > 0)
  {
    int temp = prevFib + fib;
    prevFib = fib;
    fib = temp;
  }
  return fib;
}

int fibFormula(int n)
{
  return round(pow((pow(5, .5) + 1) / 2, n) / pow(5, .5));
}

int fibTailRecursive(int n, int prevFib = 0, int fib = 1)
{
  return n == 0 ? prevFib : fibTailRecursive(n - 1, fib, prevFib + fib);
}

int* fibFastAuxiliary(int n) {
  if (n == 0) return new int[2]{0, 1};
  int* fibs = fibFastAuxiliary(floor(n / 2));
  int prevFib = fibs[0];
  int fib = fibs[1];
  int a = prevFib * (fib * 2 - prevFib);
  int b = prevFib * prevFib + fib * fib;
  return n % 2 == 0 ? new int[2]{a, b} : new int[2]{b, a + b};
}

int fibFastDoubling(int n) {
  return fibFastAuxiliary(n)[0];
}

int main(int argc, char** argv)
{
  int input = argc > 1 ? atoi(argv[1]) : 29;
  std::cout << fibFastDoubling(input) << std::endl;
}
